### PR TITLE
(For Micah) Adds Support for decisions API

### PR DESCRIPTION
**Pull request by KaoruDev**
## Purpose

Add support for new decisions API.

## Testing

- [x] Ensure we can get a list of decisions and apply filtering and paging
- [x] Ensure we can apply decisions to order / user
- [x] Ensure we properly handle api errors

## Deploy

- Merge this branch
- Bump Version to `2.1.0.0` from `2.0.0.0`
- Release to RubyGems.org

## Rollback
- Revert this branch / fix issue
- Bump gem version to `2.1.1.0` from `2.1.0.0`
- Release to RubyGems.org

### CODE REVIEW

File: lib/sift/errors.rb [hunk 1]
```diff
@@ -0,0 +1,17 @@
+module Sift
+  # Top level Error
+  class Error < StandardError
+  end
+
+  class ApiError < Error
```
Comment (KaoruDev) [inline lib/sift/errors.rb]: these are no longer used

File: spec/unit/validate/decision_spec.rb [hunk 2]
```diff
@@ -0,0 +1,67 @@
+require_relative "../../spec_helper"
+
+require "sift/validate/decision"
+
+module Sift
+  module Validate
+    describe Decision do
+      describe "#valid_user?" do
+        it "will return true" do
```
Comment (KaoruDev) [inline spec/unit/validate/decision_spec.rb]: tis done

File: lib/sift/client/decision.rb [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+      def index_path
+        "#{Client::API3_ENDPOINT}/v3/accounts/" +
+          "#{CGI.escape(account_id)}/decisions"
+      end
+
+      private
+
+      def request_next_page(path)
+        Router.get(path, query: build_query(Utils::HashGetter.new({})))
```
Comment (KaoruDev) [inline lib/sift/client/decision.rb]: don't need to build a query now that `api_key` is in `Auth` header.

File: lib/sift/client/decision/apply_to.rb [hunk 4]
```diff
@@ -0,0 +1,106 @@
+require "multi_json"
+
+require_relative "../../validate/primitive"
+require_relative "../../client"
+require_relative "../../router"
```
Comment (KaoruDev) [inline lib/sift/client/decision/apply_to.rb]: need to require utils.

File: lib/sift/client/decision/apply_to.rb [hunk 5]
```diff
@@ -0,0 +1,106 @@
+require "multi_json"
+
+require_relative "../../validate/primitive"
```
Comment (KaoruDev) [inline lib/sift/client/decision/apply_to.rb]: wrong require, `validate/decision`

File: lib/sift/client/decision/apply_to.rb [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+          "#{Client::API3_ENDPOINT}/v3/accounts/#{account_id}"
+        end
+
+        def headers
+          {
+            "Content-type" => "application/json"
+          }.merge(Client.build_auth_header(api_key))
+        end
+
+        def auth_header
```
Comment (KaoruDev) [inline lib/sift/client/decision/apply_to.rb]: remove

File: README.md [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+
+```
+gem install sift
+```
+
+## Usage
+
+### Creating a Client:
+
+```
```
Comment (mwylde) [inline README.md]: If you add ruby here these will render with syntax highlighting in github

File: lib/sift/client.rb [hunk 8]
```diff
@@ -1,17 +1,22 @@
 require 'httparty'
 require 'multi_json'
 
+require_relative "./client/decision"
```
Comment (mwylde) [inline lib/sift/client.rb]: require_relative was added in 1.9.2 — are we trying to maintain 1.8.7 compatibility?

File: lib/sift/client/decision.rb [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+      def apply_to(configs = {})
+        getter = Utils::HashGetter.new(configs)
+        configs[:account_id] = account_id
+
+        ApplyTo.new(api_key, getter.get(:decision_id), configs).run
+      end
+
+      def index_path
+        "#{Client::API3_ENDPOINT}/v3/accounts/" +
+          "#{CGI.escape(account_id)}/decisions"
```
Comment (mwylde) [inline lib/sift/client/decision.rb]: escaping shouldn't ever be necessary for an account id—these are just hex strings. Maybe it makes sense to validate those somehow?

File: lib/sift/client.rb (hunk 8 above)
Comment (KaoruDev) [inline lib/sift/client.rb]: oh i didn't realize that, I'll ask the `SE`, one thing to note is that our `travis` configs don't test `1.8.7` good catch 👍

Comment (mwylde) [inline lib/sift/client.rb]: 1.8.7 is ancient — I think dropping compatibility is fine, but we'd want to update the readme as well

Comment (KaoruDev) [inline lib/sift/client.rb]: 👍  fo sho

File: lib/sift/client/decision.rb (hunk 9 above)
Comment (KaoruDev) [inline lib/sift/client/decision.rb]: I'm gonna remove the CGI escape, but I'm gonna leave the validation up the server, that way we don't have to keep validation in sync between server and client.

**Pull request merged**